### PR TITLE
refactoring Organization classes

### PR DESCRIPTION
**Pull request by najlaalhaj**

### CODE REVIEW

**Comment by najlaalhaj** [comment]
Refactoring OrganizationAPI,OrganizationService,Organization and Database classes

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 1]
```diff
@@ -3,25 +3,61 @@
 import io.orten.nano.impl.OrganizationService;
 import io.orten.nano.model.Organization;
 import javax.ws.rs.*;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 
+// end point for all Organization operations
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: Change to use javadoc comment style, i.e.
/**
  \* documentation
  */

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 2]
```diff
@@ -3,25 +3,61 @@
...
+
 @Path("/org")
 public class OrganizationAPI {
 
+    //a method to save an organization object in the database
+
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: use javadoc style. se other comment

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 3]
```diff
@@ -3,25 +3,61 @@
...
-    @Path("/saveOrg")
+    @Path("/saveorg")
     @Consumes(MediaType.APPLICATION_JSON)
-    public Response saveInDatabase(Organization org)
+    public Response saveOrganization(Organization org)
     {
         return new OrganizationService().save(org);
     }
 
+    //a method to update an organization object that already saved in the database
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: use javadoc style. se other comment

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 4]
```diff
@@ -3,25 +3,61 @@
...
-    @Path("/updateOrg")
+    @Path("/updateorg")
     @Consumes(MediaType.APPLICATION_JSON)
     public Response updateOrg(Organization org)
     {
-      return new OrganizationService().update(org);
+        return new OrganizationService().update(org);
+    }
+  
+    //a method to retrieve an organization object from the database based on the organizationID
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: use javadoc style. se other comment

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 5]
```diff
@@ -3,25 +3,61 @@
...
+        return new OrganizationService().get(orgID);
+    }
+
+    //a method to retrieve allorgnizations objects that are  already saved in the database
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("/getall")
+    public Response getAllOrg(){
+        return new io.orten.nano.impl.OrganizationService().getAll();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: remove full package name as it is not needed here

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 6]
```diff
@@ -3,25 +3,61 @@
...
     }
-}
\ No newline at end of file
+
+    //a method to delete an organization object from the database
+    @DELETE
+    @Consumes
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: This annotation is not doing anything

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 7]
```diff
@@ -3,25 +3,61 @@
...
+    @Path("/deleteorg")
+    @Produces(MediaType.TEXT_PLAIN)
+    public Response deleteOrg(@QueryParam("orgID") String orgID){
+        return new io.orten.nano.impl.OrganizationService().delete(orgID);
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: Remove full package name

File: src/main/java/io/orten/nano/business/OrganizationAPI.java [hunk 8]
```diff
@@ -3,25 +3,61 @@
...
+        return new io.orten.nano.impl.OrganizationService().getAll();
     }
-}
\ No newline at end of file
+
+    //a method to delete an organization object from the database
+    @DELETE
+    @Consumes
+    @Path("/deleteorg")
+    @Produces(MediaType.TEXT_PLAIN)
```
Comment (munHunger) [inline src/main/java/io/orten/nano/business/OrganizationAPI.java]: Why are you suddenly producing TEXT_PLAIN?

File: src/main/java/io/orten/nano/impl/OrganizationService.java [hunk 9]
```diff
@@ -6,25 +6,60 @@
...
+        }
+    }
+
+    public Response get(String orgID)
+    {
+        try {
+            return Response.status(HttpServletResponse.SC_FOUND).entity(Database.getOrganization(orgID)).build();
+        }
+        catch(Exception e) {
+            return Response.status(HttpServletResponse.SC_NOT_FOUND).entity(e.getMessage()).build();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/OrganizationService.java]: This is bad practice.
What would happen if for example the database is unavailable?
I would recomend checking if the database returns null, and if so, then return 404.

File: src/main/java/io/orten/nano/impl/OrganizationService.java [hunk 10]
```diff
@@ -6,25 +6,60 @@
...
+        }
+    }
+
+    public Response getAll() {
+        try {
+            return Response.status(HttpServletResponse.SC_FOUND).entity(Database.getAllOrganizations()).build();
+        } catch (Exception e) {
+            return Response.status(HttpServletResponse.SC_NO_CONTENT).entity(e.getMessage()).build();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/OrganizationService.java]: Same as for get.
In 90% of all cases you should only return a 500 server error if you catch an exception

File: src/main/java/io/orten/nano/impl/OrganizationService.java [hunk 11]
```diff
@@ -6,25 +6,60 @@
...
+        }
+    }
+
+    public Response delete(String orgID){
+        try{
+            return Response.status(HttpServletResponse.SC_GONE).entity(Database.deleteOrganization(orgID)).build();
+        }
+        catch(Exception e){
+            return Response.status(HttpServletResponse.SC_NOT_FOUND).entity(e.getMessage()).build();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/OrganizationService.java]: Same as for get

File: src/main/java/io/orten/nano/impl/OrganizationService.java [hunk 12]
```diff
@@ -6,25 +6,60 @@
...
+        try {
+            return Response.status(HttpServletResponse.SC_FOUND).entity(Database.getAllOrganizations()).build();
+        } catch (Exception e) {
+            return Response.status(HttpServletResponse.SC_NO_CONTENT).entity(e.getMessage()).build();
+        }
+    }
+
+    public Response delete(String orgID){
+        try{
+            return Response.status(HttpServletResponse.SC_GONE).entity(Database.deleteOrganization(orgID)).build();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/OrganizationService.java]: SC_GONE is an error code and thus shouldn't be returned if everything went as expected

File: src/main/java/io/orten/nano/impl/OrganizationService.java [hunk 13]
```diff
@@ -6,25 +6,60 @@
...
+        }
+        catch(Exception e){
+            return Response.status(HttpServletResponse.SC_NOT_FOUND).entity(e.getMessage()).build();
+        }
+    }
+}
+
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/OrganizationService.java]: Remove empty lines

File: src/main/java/io/orten/nano/model/Organization.java [hunk 14]
```diff
@@ -1,42 +1,22 @@
...
-import org.hibernate.annotations.DynamicUpdate;
-import org.hibernate.annotations.SelectBeforeUpdate;
 import javax.persistence.*;
-import java.io.Serializable;
-
-/**
-* represents an organization raising fund for one or more projects
-*/
+/*
+  represents an organization raising fund for one or more projects
```
Comment (munHunger) [inline src/main/java/io/orten/nano/model/Organization.java]: use javadoc code style

File: src/main/java/io/orten/nano/model/Organization.java [hunk 15]
```diff
@@ -45,20 +25,20 @@ public Organization()
     {
 
     }
-    /**
+
+ /**
```
Comment (munHunger) [inline src/main/java/io/orten/nano/model/Organization.java]: fix indentation

File: src/main/java/io/orten/nano/util/Database.java [hunk 16]
```diff
@@ -1,64 +1,113 @@
...
 import org.hibernate.SessionFactory;
 import org.hibernate.boot.Metadata;
 import org.hibernate.boot.MetadataSources;
 import org.hibernate.boot.registry.StandardServiceRegistry;
 import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
+import org.hibernate.query.Query;
+import javax.xml.ws.Response;
+import java.util.List;
+
+//  Creates and manages connections with and transactions to the database
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: use javadoc code style

File: src/main/java/io/orten/nano/util/Database.java [hunk 17]
```diff
@@ -1,64 +1,113 @@
...
 
-        public static void updateOrganization(Object o){
-            if(sessionFactory == null)
-                init();
-            try(Session s = sessionFactory.openSession()){
-                s.beginTransaction();
-                s.update(o);
-                s.getTransaction().commit();
-            }
+    public static List<Organization> getOrganization(String orgID) {
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: It is quite wierd to return a list of one Organization.
You should probably return the first one if one exists(or null).

File: src/main/java/io/orten/nano/util/Database.java [hunk 18]
```diff
@@ -1,64 +1,113 @@
...
+            List list = q.getResultList();
+            s.getTransaction().commit();
+            return list;
         }
+    }
 
-        /**
-        * TODO: This need to be refactor
-        */
+    public static String deleteOrganization(String orgID){
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: Why return a String from the delete?
Could probably return a boolean for if something to delete was found or not

File: src/main/java/io/orten/nano/util/Database.java [hunk 19]
```diff
@@ -1,64 +1,113 @@
...
-            try(Session s = sessionFactory.openSession()){
-                s.beginTransaction();
-                s.update(o);
-                s.getTransaction().commit();
-            }
+    public static List<Organization> getOrganization(String orgID) {
+        if (sessionFactory == null)
+            init();
+        try (Session s = sessionFactory.openSession()) {
+            s.beginTransaction();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: You don't need to begin(or commit) a transaction if you are just retrieving data

File: src/main/java/io/orten/nano/util/Database.java [hunk 20]
```diff
@@ -1,64 +1,113 @@
...
+        }      
+     }
+      
+    //gets all the organization objects from the database
+
+    public static List<Organization> getAllOrganizations(){
+        if (sessionFactory == null)
+            init();
+        try (Session s = sessionFactory.openSession()) {
+            s.beginTransaction();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: You don't need to begin(or commit) a transaction if you are just retrieving data

File: src/main/java/io/orten/nano/util/Database.java [hunk 21]
```diff
@@ -1,64 +1,113 @@
...
+        }
+    }
 
-        public static Session getSession() throws Exception {
-            if(sessionFactory == null) {
-                init();
-            }
-            Session session = sessionFactory.openSession();
-            return session;
+    public static Session getSession() throws Exception {
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: Is this used?

File: src/main/java/io/orten/nano/util/Database.java [hunk 22]
```diff
@@ -1,64 +1,113 @@
...
+        if(sessionFactory == null) {
+            init();
         }
- }
\ No newline at end of file
+        Session session = sessionFactory.openSession();
+        return session;
+    }
+}
+
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java:156]: Remove empty lines

**Review by munHunger — CHANGES_REQUESTED** [review]

File: src/main/java/io/orten/nano/impl/OrganizationService.java [hunk 23]
```diff
@@ -4,27 +4,80 @@
...
+                return Response.status(HttpServletResponse.SC_OK).build();
+            }
+            else {return Response.status(HttpServletResponse.SC_NOT_FOUND).build();}
+        }
+        catch(Exception e){
+            return Response.status(HttpServletResponse.SC_INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
+        }
+    }
+}
+
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/OrganizationService.java:77]: lot of empty lines here

File: src/main/java/io/orten/nano/impl/ProjectService.java [hunk 24]
```diff
@@ -66,4 +70,45 @@ public static void deleteProject(String projectID) throws Exception {
...
+    public static List<Project> getFundedProjects(String donorID) throws Exception{
+        List<Project> fundedProjects = new ArrayList<>();
+        Session session = Database.getSession();
+        Transaction tx= null;
+        try{
+            tx =session.beginTransaction();
+            Query query = session.createQuery("from Project where donor.userID= :donorID");
+            query.setParameter("donorID",donorID);
+            fundedProjects=query.getResultList();
+            if (fundedProjects.size()!=0){
```
Comment (munHunger) [inline src/main/java/io/orten/nano/impl/ProjectService.java]: Could be replaced with fundedProjects.isEmpty()

File: src/main/java/io/orten/nano/util/Database.java [hunk 25]
```diff
@@ -1,64 +1,161 @@
...
-                s.getTransaction().commit();
+    /**
+     * updates an organization object already saved in saved in the database
+     */
+    public static void updateOrganization(Organization org) {
+        if (sessionFactory == null)
+            init();
+        try (Session s = sessionFactory.openSession()) {
+            s.beginTransaction();
+            s.update(org);
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java:68]: shouldn't you also commit the transaction?

File: src/main/java/io/orten/nano/util/Database.java [hunk 26]
```diff
@@ -1,64 +1,161 @@
...
+     * saves a donor record in the database
+     */
+    public static boolean saveUser(User user) {
+        if (sessionFactory == null)
+            init();
+        try (Session s = sessionFactory.openSession()) {
+            s.getTransaction();
+            if (user.getUserID() != null) {
+                s.save(user);
+                s.beginTransaction().commit();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: begin a transaction and the commit it right away?
that seems odd

File: src/main/java/io/orten/nano/util/Database.java [hunk 27]
```diff
@@ -1,64 +1,161 @@
...
+
+    /**
+     * gets one organization object from the database based on its ID
+     */
+    public static Organization getOrganization(String orgID) {
+        if (sessionFactory == null)
+            init();
+        try (Session s = sessionFactory.openSession()) {
+            List<Organization> list = new ArrayList<>();
+            s.beginTransaction();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: Why begin a transaction if you are not going to commit it?
in this case you don't need to begin a transaction

File: src/main/java/io/orten/nano/util/Database.java [hunk 28]
```diff
@@ -1,64 +1,161 @@
...
 
-        /**
-         * updating an organization object
-        */
+    public static List<Organization> getAllOrganizations(){
+        List<Organization> list= new ArrayList<>();
+        if (sessionFactory == null)
+            init();
+        try (Session s = sessionFactory.openSession()) {
+            s.beginTransaction();
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: Why begin a transaction if you are not going to commit it?
in this case you don't need to begin a transaction

File: src/main/java/io/orten/nano/util/Database.java [hunk 29]
```diff
@@ -1,64 +1,161 @@
...
+            } else {
+                return false;
             }
-            Session session = sessionFactory.openSession();
-            return session;
         }
- }
\ No newline at end of file
+    }
+
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java:144]: empty lines that could be removed

File: src/main/java/io/orten/nano/util/Database.java [hunk 30]
```diff
@@ -1,64 +1,161 @@
...
+     * belongs to ProjectService class
+     */
+    public static Session getSession() throws Exception {
+        if(sessionFactory == null) {
+            init();
+        }
+        Session session = sessionFactory.openSession();
+        return session;
+    }
+
```
Comment (munHunger) [inline src/main/java/io/orten/nano/util/Database.java]: empty lines that could be removed

**Review by munHunger — CHANGES_REQUESTED** [review]

**Review by munHunger — APPROVED** [review]

**Comment by najlaalhaj** [comment]
finally approved!

**Pull request merged**